### PR TITLE
Example Flamethrower

### PR DESCRIPTION
**Pull request by YuriPrickles**
The **Example Flamethrower** is a cool (hot) way for new users to learn about flamethower-like weapons like the Flamethrower and Elf Melter. These weapons have unique properties than other ranged weapons like guns and bows.

This pull request includes:
**ExampleFlamethrower.cs
FlamethrowerProj.cs** and its corresponding texture (the texture doesnt really matter since it gets set to be invisible)

### CODE REVIEW

File: ExampleMod/Items/Weapons/ExampleFlamethrower.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+using Microsoft.Xna.Framework;
+using Terraria;
+using Terraria.ID;
+using Terraria.ModLoader;
+using ExampleMod.Projectiles;
+using static Terraria.ModLoader.ModContent;
```
Comment (Steviegt6) [inline ExampleMod/Items/Weapons/ExampleFlamethrower.cs:6]: Don't use static directives.

File: ExampleMod/Items/Weapons/ExampleFlamethrower.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+			item.useAnimation = 20; //Shoots out 5 jets of fire in one shot
+			item.useStyle = ItemUseStyleID.HoldingOut;
+			item.noMelee = true; //So the item's animation doesn't do damage
+			item.knockBack = 2;
+			item.value = 10000;
+			item.color = new Color(61, 252, 3); //Makes the item color green.
+			item.rare = ItemRarityID.Green; //Sets the item's rarity.
+			item.UseSound = SoundID.Item34;
+			item.autoReuse = true;
+            item.shoot = ProjectileType<FlamethrowerProj>();
```
Comment (Steviegt6) [inline ExampleMod/Items/Weapons/ExampleFlamethrower.cs:35]: Fix your indentation.

File: ExampleMod/Items/Weapons/ExampleFlamethrower.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			item.shootSpeed = 21f; //How fast the flames will travel
+			item.useAmmo = AmmoID.Gel; //Makes the weapon use up Gel as ammo
+		}
+		public override bool ConsumeAmmo(Player player)
```
Comment (Steviegt6) [inline ExampleMod/Items/Weapons/ExampleFlamethrower.cs:39]: Specify what ths actually does.

File: ExampleMod/Items/Weapons/ExampleFlamethrower.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			recipe.AddIngredient(ItemType<ExampleItem>(), 3);
+			recipe.AddIngredient(ItemID.CursedFlame, 15);
+			recipe.AddTile(TileType<Tiles.ExampleWorkbench>());
+			recipe.SetResult(this);
+			recipe.AddRecipe();
+		}
+
+        public override bool Shoot(Player player, ref Vector2 position, ref float speedX, ref float speedY, ref int type, ref int damage, ref float knockBack)
+        {
+			Vector2 muzzleOffset = Vector2.Normalize(new Vector2(speedX, speedY)) * 54f;
```
Comment (Steviegt6) [inline ExampleMod/Items/Weapons/ExampleFlamethrower.cs:56]: Specify what this actually does.

File: ExampleMod/Items/Weapons/ExampleFlamethrower.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			if (Collision.CanHit(position, 0, 0, position + muzzleOffset, 0, 0))
+			{
+				position += muzzleOffset;
+			}
+			//This is to prevent shooting through blocks and to make the fire shoot from the muzzle.
+			return true;
+        }
+        public override Vector2? HoldoutOffset()
```
Comment (Steviegt6) [inline ExampleMod/Items/Weapons/ExampleFlamethrower.cs:64]: More indentation issues + explain what the vector is actually for.

**Review by steviegt6 — CHANGES_REQUESTED** [review]

File: ExampleMod/Projectiles/FlamethrowerProj.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+			projectile.timeLeft = 30; //A short life time for this projectile to get the flamethrower effect
+			projectile.ignoreWater = false;
+			projectile.tileCollide = true;
+		}
+
+		public override void AI()
+		{
+			/*if (projectile.wet)
+			{
+				projectile.Kill(); //This kills the projectile when touching water. However, since our projectile is a cursed flame, we will comment this so that it won't run it. If you want to test this, feel free to uncomment this lol
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/FlamethrowerProj.cs:33]: The "lol" at the end is kinda weird

**Review by steviegt6 — CHANGES_REQUESTED** [review]

File: ExampleMod/Items/Weapons/ExampleFlamethrower.cs (hunk 2 above)
Comment (RighteousRyan1) [reply ExampleMod/Items/Weapons/ExampleFlamethrower.cs:35]: Highly certain this is a GitHub artifact, for some reason it ruins indentation.

Comment (Steviegt6) [reply ExampleMod/Items/Weapons/ExampleFlamethrower.cs:35]: GitHub doesn't cause this kind of issue. Inconsistencies in tabbing do (i.e. mixed spaces and tabs).

**Comment by JavidPack** [comment]
Thanks a lot

File: ExampleMod/Projectiles/FlamethrowerProj.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+			}*/
+			for (int k = 0; k < 10; k++) 
+			{
+				if (Main.rand.Next(5) == 0) 
+				{
+					Vector2 position = projectile.Center;
+					Dust dust = Main.dust[Terraria.Dust.NewDust(position, 16, 16, 75, 0f, 0f, 140, new Color(61, 252, 3), Main.rand.NextFloat(2.6f, 4.4f))];
```
Comment (timoha40k) [inline ExampleMod/Projectiles/FlamethrowerProj.cs]: How to change color of dust ?

**Pull request merged**